[PLATFORM-1483] User permissions lib

jcellary opened the pull request:
This PR is centralizing permission handling inside of media wiki, so that later on we can move that logic to the user permissions service.

For that purpose a new permission service lib has been introduced (similarly to user preferences) and all logic from the User class is moved to the lib. Methods in the User class now only serve as a proxy. 

From an altered lines perspective most of changes are moving permission definition to a single file located in config repo.

This PR needs to be merged together with https://github.com/Wikia/config/pull/1620

Commit:
Created Permissions Service

Commit:
User Permissions refactoring - removed parts of autopromote

Commit:
Moved getImplicitGroups

Commit:
Removed wgAvailableRights and put all definitions together

Commit:
Moved Shared User to PermissionService

Commit:
Fixes in content-reviewer handling and others

Commit:
Moved all global group permissions

Commit:
Converted tests to PermissionTest user

Commit:
Merged with dev

Commit:
FixeUserGetRights hook

Commit:
Cleanup in comments

Commit:
Merge branch 'user-permissions-lib' of github.com:Wikia/app into user-permissions-lib

Commit:
Removed internal and fixes permission query

Commit:
Fixed missing import

Commit:
Removed accidentally added LicensedVideoSwap

Commit:
Removed accidentally added Phalanx

macbre commented on extensions/wikia/SpecialBatchUserRights/SpecialBatchUserRights.php:
```diff
@@ -42,6 +49,17 @@ public function __construct() {
+
+		return $this->permissionsService;
+	}
+
```
This one is repeated in quite a few places. Let's consider introducing a trait that will provide `userPermissions` method.

wladekb commented on extensions/3rdparty/LyricWiki/Hook_PreventBlanking.php:
```diff
@@ -3,8 +3,6 @@
 $wgHooks['EditFilter'][] = 'stopBlanking';
-$wgAvailableRights[] = 'allowedtoblank';
-$wgGroupPermissions['sysop']['allowedtoblank'] = true;
 $wgExtensionMessagesFiles["PreventBlanking"] = "$LW/Hook_PreventBlanking.i18n.php";
```
If you are removing the rights, is this still in use?

jcellary replied:
Not sure I understand - I'm not removing any rights, I'm just moving all definitions to one single place in config.

wladekb replied:
yeah, makes sense, nvm

wladekb commented on extensions/3rdparty/LyricWiki/LyricFind/LyricFind.setup.php:
```diff
@@ -31,12 +31,6 @@
-$wgGroupPermissions['staff']['editlyricfind'] = true;
-$wgGroupPermissions['sysop']['editlyricfind'] = true;
-
 $wgHooks['AlternateEdit'][] = 'LyricFindHooks::onAlternateEdit';
```
Please find out if we can remove everything related to LyricFind. AFAIK we stopped using that service X years ago.
@owend @SeanColombo

SeanColombo replied:
@wladekb nope! We stopped using Gracenote, but we VERY MUCH use LyricFind curreclty and that's how we stay licensed.

I know it's confusing because it changes between a bunch of things with similar names, but LyricFind _is_ our current sub-licensor :)

wladekb replied:
@SeanColombo Thanks for confirmation!

wladekb commented on extensions/wikia/ContentReview/ContentReview.hooks.php:
```diff
@@ -341,26 +338,6 @@ public function onShowDiff( \DifferenceEngine $diff, &$notice ) {
-	public function onUserAddGroup( \User $user, $group ) {
-		global $wgUser;
-
-		if ( $group === 'content-reviewer' && ( !$wgUser->isAllowed( 'content-review' ) || !$wgUser->isStaff() ) ) {
```
where this logic will be placed now?

jcellary replied:
onUserAddGroup is part of a check inside of the user permissions service. As for GroupCheckboxes I moved that logic directly to the special page displaying it

wladekb replied:
Please give a direct link to the relevant part of special page code, this PR is big!

jcellary replied:
So for this particular check, I talked to Adam and we agreed it's sufficient to specify that only members of content-reviewer group are allowed to assign this group (and skip the is staff check). So this is part of this definition:
https://github.com/Wikia/app/pull/9720/files#diff-8bf819f58bde706eaeeb9261498f35e5R384

wladekb commented on extensions/wikia/GameGuides/GameGuides_setup.php:
```diff
@@ -42,16 +38,7 @@
-$wgGroupPermissions['helper']['gameguidescontent'] = true;
-
-if ( $wgGameGuidesContentForAdmins ) {
-	$wgGroupPermissions['sysop']['gameguidescontent'] = true;
```
how are we going to support this kind of logic?

jcellary replied:
I moved these lines as is to the new place, so the if will stay as is. You're right however that I can double check when is the wgGameGuidesContentForAdmins set to make sure we don't load rights before the value of this variable is set.

wladekb replied:
thanks for explanation, should be fine

wladekb commented on extensions/wikia/Listusers/SpecialListusers_helper.php:
```diff
@@ -22,6 +25,11 @@ class ListusersData {
+	/**
+	 * @var UserPermissions
+	 */
+	private $permissionsService;
```
`$userPermissionsService` ? so it will be consistent with the accessor method name `userPermissions` ?

jcellary replied:
True. I'll convert it to a trait as Macbre suggested

wladekb commented on extensions/wikia/PowerUser/PowerUser.class.php:
```diff
@@ -69,6 +71,22 @@ function __construct( \User $oUser ) {
 	/**
+	 * @var UserPermissions
+	 */
+	private $permissionsService;
```
`$userPermissionsService` ? (and all the other classes too)

wladekb replied:
Any answer?

jcellary replied:
The answer is that I already converted it to the trait that you suggested :)

wladekb replied:
That particular question was about the name of the variable...

jcellary replied:
Ok, this code is deleted now and move to the trait, so I'm assuming the issue is resolved

wladekb commented on extensions/wikia/PowerUser/PowerUser.class.php:
```diff
@@ -122,10 +140,10 @@ public function addPowerUserSetOption( $sProperty ) {
 		if ( in_array( $sProperty, self::$aPowerUserProperties )
 			&& $this->bUseGroups
-			&& !in_array( self::GROUP_NAME, \UserRights::getGlobalGroups( $this->oUser ) )
+			&& !in_array( self::GROUP_NAME, $this->userPermissions()->getExplicitGlobalUserGroups( $this->oUser->getId() ) )
```
it would be much more readable if we could write something like:
`$this->userPermissions()->hasGlobalGroup( $this->oUser->getId(), self::GROUP_NAME )`
or:
`$this->userPermissions()->user( $this->oUser )->hasGlobalGroup( self::GROUP_NAME )` (or `forUser` or `withUser` or... ) - it would be convenient if such a method would accept both integer IDs and MW's User objects (or at least we had a chance to pass a User object to a separate method)

can we add a handful of helper methods to the user permissions service class and use them wherever possible?

-and-

we could have two private methods:
- `userPermissionsService()` that returns UserPermissionsService object
- `userPermissions( $user )` that returns a proxy to UserPermissionsService bound to a specific user (that would shorten constructs like this one: `$this->userPermissions()->user( $this->oUser )`

wladekb replied:
Any answer?

jcellary replied:
I made the suggested change of unifying parameters - now all methods get the user objects instead of user id/user object. As for the proxy you suggested, I considered that, however part of methods take the user object for which we're performing a check, part take a user on which the operation is being performed and some both, so It wouldn't be possibly to do straightforward binding of the whole API to a single user object as its interpretation depends on the function context.

wladekb commented on extensions/wikia/PowerUser/PowerUser.class.php:
```diff
@@ -69,6 +71,22 @@ function __construct( \User $oUser ) {
+	/**
+	 * @return UserPermissions
+	 */
+	private function userPermissions() {
```
this method and the private field is repeated unchanged in many classes, good candidate for `trait` ?

macbre replied:
https://github.com/Wikia/app/pull/9720#discussion_r53135072 :)

wladekb commented on includes/Autopromote.php:
```diff
@@ -28,45 +28,6 @@ public static function getAutopromoteGroups( User $user ) {
 	}
 
 	/**
-	 * Get the groups for the given user based on the given criteria.
```
will we use Autopromote for anything after this PR is merged? have you considered removing everything related to autopromote?

jcellary replied:
Yes, autopromote per se is still used. What is not used is AutopromoteONCE - that is a subset of the whole autopromote logic.

macbre commented:
One general suggestion: this pull request is still in progress, but already quite big. Can we perform a cleanup of code
- Removal of unused extension 3rdparty/SpecialProtectSite (not to confuse with wikia/SpecialProtectSite)
- Removal of `WIKIA_ENV_INTERNAL` (retired wikia one)

as a part of a separate pull request to make things a bit simpler and release these (quite big) changes in batches?

wladekb commented on includes/User.php:
```diff
@@ -5237,4 +4613,256 @@ protected function logFallbackToMediaWikiSessionRejection( $from ) {
+	 * @param $group String Name of the group to add
+	 */
+	public function addGroup( $group ) {
+		global $wgUser;
```
why are you changing a behavior here?
the original method operated on the current object, not on the currently logged-in user

the same below in a few places

wladekb commented on includes/User.php:
```diff
@@ -5237,4 +4613,256 @@ protected function logFallbackToMediaWikiSessionRejection( $from ) {
+	 */
+	public function addGroup( $group ) {
+		global $wgUser;
+		return self::userPermissions()->addUserToGroup( $wgUser, $this, $group );
```
use `RequestContext::getMain()->getUser()` instead of `$wgUser`
the same below

wladekb replied:
That one hasn't been addressed yet.

jcellary replied:
Ok, I can change that, but you could explain why using wgUser is bad? Just because we want to avoid globals in general is there some timing issue? I'm asking because I use wgUser in other places as well

wladekb commented on includes/specials/SpecialListgrouprights.php:
```diff
@@ -113,11 +125,13 @@ public function execute( $par ) {
-			$removegroups = isset( $wgRemoveGroups[$group] ) ? $wgRemoveGroups[$group] : array();
-			$addgroupsSelf = isset( $wgGroupsAddToSelf[$group] ) ? $wgGroupsAddToSelf[$group] : array();
-			$removegroupsSelf = isset( $wgGroupsRemoveFromSelf[$group] ) ? $wgGroupsRemoveFromSelf[$group] : array();
+			$addgroups = isset( $groupArr['add'] ) ? $groupArr['add'] : array();
```
how about enforcing `getGroupsChangeableByGroup` to return all those keys and then we could skip checking if they exist here?

wladekb replied:
Any answer?

jcellary replied:
Good point, actually this is already enforced, so I can just remove the checks

wladekb commented on includes/specials/SpecialStatistics.php:
```diff
@@ -223,8 +243,7 @@ private function getGroupStats() {
 				// wikia change start
-				global $wgWikiaGlobalUserGroups;
-				if( is_array( $wgWikiaGlobalUserGroups ) && in_array( $groupname, $wgWikiaGlobalUserGroups ) ) {
+				if( in_array( $groupname, $this->userPermissions()->getGlobalGroups() ) ) {
```
the helper methods would be quite nice here and there...

wladekb replied:
Any answer?

jcellary replied:
Yes, the answer is no.

...ok I replaced all in_array( getXXXgroups )

wladekb commented on includes/specials/SpecialUserrights.php:
```diff
@@ -504,6 +511,11 @@ protected static function getAllGroups() {
 		return User::getAllGroups();
 	}
 
+	private function isGlobalNonEditableGroup( $group ) {
```
I've seen this before, worth moving it to a shared place?

jcellary replied:
moved

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsService.php:
```diff
@@ -0,0 +1,43 @@
+namespace Wikia\Service\User\Permissions;
+
+interface PermissionsService {
+	public function getGlobalGroups();
```
would be quite helpful to have phpdocs describing parameter types and return values

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+	}
+
+	private function loadGlobalGroups() {
+		$this->globalGroups[] = 'content-reviewer';
```
```
$this->globalGroups = [
   'one',
   'two',
   ....
```

or even move it to the property declaration

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+	}
+
+	private function loadImplicitGroups() {
+		$this->implicitGroups[] = '*';
```
same here as above

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+			'abusefilter-bypass',
+			'platinum',
+			'sponsored-achievements',
+			'achievements-exempt',
```
achievements are not enabled everywhere and this right should probably not be listed on wikis that don't have it

jcellary replied:
That's a case with a lot of above mentioned permissions, but that was one of intended changes - that we make permissions definition global not extension based. When moving the logic to the external service there will be no notion of extensions so if you're in a specfic group you simply get all rights defined for it.

wladekb replied:
IMHO in such case we need a single source of truth (simple directory?) for the permission names so we can assure there will be no clash in the future when someone adds a new permission not knowing that such one already exists. Makes sense?

jcellary replied:
The single source of truth is this file. In future this file will be part of the user permissions service.

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+			'commentedit',
+			'commentdelete',
+			'becp_user',
+			'blog-comments-toggle',
```
same for blogs etc.

jcellary replied:
same comment as above

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsService.php:
```diff
@@ -0,0 +1,43 @@
+
+	public function getPermissions();
+
+	public function getExplicitUserGroups( $userId );
```
why some of the methods accept IDs and some others User objects?

jcellary replied:
Because some function need only the id, however some other use methods defined in the user class. But you're right it can be confusing - I'll take a second look if we can standardize it, till now I simply based it on actual needs.

jcellary commented:
You're right Macbre, I'll create a separate PR for the cleanups and release it step by step.

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+
+	public function getExplicitLocalUserGroups( $userId ) {
+		$this->loadLocalGroups( $userId );
+		return $this->getLocalUserGroupsArray( $userId );
```
I'd suggest naming those methods in such a way it's more obvious which ones load the data and which one don't
I wouldn't guess which on of them does: `getExplicitLocalUserGroups` or `getLocalUserGroupsArray`

jcellary replied:
done, these are deleted now

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+			return [];
+		}
+
+		return array_unique( array_merge (
```
calculated every time....

wladekb replied:
Any answer?

jcellary replied:
The underlying logic caching its result the only thing that is recalculated is array unique/merge. Does it really affect performance that much?

Btw. I believe I answered this comment in another spot, but can't find it now...

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+	 */
+	public function getEffectiveUserGroups( \User $user, $reCacheAutomaticGroups = false ) {
+
+		return array_unique( array_merge(
```
calculated every time... (please also look at other methods when fixing it)

wladekb replied:
Any answer?

jcellary replied:
see above

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+		if ( !empty( $userId ) && $this->getGlobalUserGroupsArray( $userId ) == false ) {
+
+			$fname = __METHOD__;
+			$globalGroups = \WikiaDataAccess::cache(
```
this should be also invalidated when `User::invalidateCache` is called

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+		}
+
+		// Purge cache first so in case of DB failure we don't leave inconsistent data in cache
+		\WikiaDataAccess::cachePurge( self::getMemcKey( $user->getId() ) );
```
race condition here... 
one can read data from DB between you cache purge and writing the new value to DB, especially when the data is read from slave

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+		return true;
+	}
+
+	private function canGroupBeAdded( \User $userPerformingChange, \User $userToChange, $group ) {
```
`$userPerformingChange` is a bit long, how about `$performer`, `$doer` or something else which is shorter?

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,832 @@
+			$groups = array_merge_recursive(
+				$groups, $this->getGroupsChangeableByGroup( $addergroup )
+			);
+			$groups['add']    = array_unique( $groups['add'] );
```
I guess we could do a single pass of `array_unique` calls after the `foreach` loop, is that correct?

jcellary replied:
done

Commit:
Removed unused field from User class

Commit:
Merged with master

Commit:
Merged with dev and removed wgRevokePermissions

Commit:
Refactoring based on PR comments

Commit:
Fixed unit tests

Commit:
Fixed usage of renamed functions

Commit:
Cleaned rights according to tickets from comsup

jcellary commented:
@wladekb I believe I addressed all your comments, let me know if you have anything else. Otherwise I'll run some more tests and be getting close to merging it

Commit:
Fixed usage user user permissions

Commit:
Fixed handling of arrays with user permissions

Commit:
Fixed issue in special batch user method usage

Commit:
Adjusted content review unit tests to changes in permissions

wladekb commented on extensions/wikia/PowerUser/PowerUser.class.php:
```diff
@@ -12,9 +12,13 @@
 
 class PowerUser {
 	use Loggable;
+	use PermissionsAccessor;
```
The name of the trait doesn't suggest you will access a service, how about renaming it to `PermissionsServiceAccessor` ?

jcellary replied:
renamed

wladekb commented on extensions/wikia/PowerUser/PowerUser.class.php:
```diff
@@ -122,10 +126,11 @@ public function addPowerUserSetOption( $sProperty ) {
 		if ( in_array( $sProperty, self::$aPowerUserProperties )
 			&& $this->bUseGroups
-			&& !in_array( self::GROUP_NAME, \UserRights::getGlobalGroups( $this->oUser ) )
+			&& !$this->permissionsService()->isUserInGroup( $this->oUser, self::GROUP_NAME )
```
here we have `permissionsService()` ...

jcellary replied:
missed that, thanks

Commit:
Merged with dev

wladekb commented on extensions/wikia/PowerUser/PowerUser.class.php:
```diff
@@ -174,7 +179,7 @@ public function removePowerUserRemoveGroup( $sProperty ) {
 			&& $this->isGroupForRemoval( $sProperty )
 		) {
-			\UserRights::removeGlobalGroup( $this->oUser, self::GROUP_NAME );
+			$this->userPermissions()->removeUserFromGroup( $this->oUser, self::GROUP_NAME );
```
... and here `userPermissions()` - is something different returned or it's just in the middle of rename?

jcellary replied:
done

wladekb commented on includes/specials/SpecialUserrights.php:
```diff
@@ -45,15 +106,6 @@ public function userCanExecute( User $user ) {
 		return $this->userCanChangeRights( $user, false );
```
You have converted method `userCanChangeRights` to static. Please make sure you adjust all the places where this is called to call it statically.

jcellary replied:
done

Commit:
Renamed trait and fixed power user permissions usage

wladekb commented on includes/User.php:
```diff
@@ -309,6 +246,17 @@ private function userAttributes() {
 	/**
+	 * @return PermissionsService
+	 */
+	private static function permissionsService() {
```
`use PermissionsAccessor` ?

wladekb replied:
Oh, this is static here. why is it so?

jcellary replied:
Because some methods that need it in the user class are static as well

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsAccessor.php:
```diff
@@ -0,0 +1,24 @@
+	/**
+	 * @return PermissionsService
+	 */
+	private function permissionsService() {
```
I'd suggest `protected` here. Otherwise if class `A` uses `PermissionsAccessor` and the subclass `B` will also want to do it, it might have some problems IIRC.

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsDefinition.php:
```diff
@@ -0,0 +1,55 @@
+
+namespace Wikia\Service\User\Permissions;
+
+interface PermissionsDefinition {
```
`...Configuration` ?

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsDefinitionImpl.php:
```diff
@@ -0,0 +1,414 @@
+		'commentedit',
+		'commentdelete',
+		'becp_user',
+		'blog-comments-toggle',
```
was it somehow addressed that this shouldn't be on this list on wikis that don't have blogs enabled (or you're sure we can change this behavior) ?

jcellary replied:
Yeah, that can be a problem - I want to send this version to com sup, and I explicitly wrote about this change asking if it's acceptable so we'll see what they will say.

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsService.php:
```diff
@@ -0,0 +1,122 @@
+	 * @param \User $user
+	 * @return string[] Get list of global and local groups to which the given user has been explicitly assigned to.
+	 */
+	public function getExplicitUserGroups( \User $user );
```
all the methods in this class accept a `User` object and operate on user, can we get rid of all the `User`'s in the method names?

jcellary replied:
done

Commit:
Fixed usage of userCanChangeRights

Commit:
Renamed permissions functions

jcellary commented:
@wladekb I made changes according to your comments, let me know if you have any other suggestions

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsService.php:
```diff
@@ -0,0 +1,122 @@
+	 * @param $permission string Name of permission
+	 * @return bool True if the user has the provided permission
+	 */
+	public function doesUserHavePermission( \User $user, $permission );
```
`hasPermission` would be much shorter and IMHO such a naming is popular within Mediawiki ( @macbre agree? ) ?

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsService.php:
```diff
@@ -0,0 +1,122 @@
+
+namespace Wikia\Service\User\Permissions;
+
+interface PermissionsService {
```
I can see you didn't reply to one of my suggestions to introduce a "bound PermissionsService" using:
`$this->permissionsService()->user( $someUserObject )->_aMethodFromThisInterface_()`

Could you share your point of view on this topic?

jcellary replied:
I answered to that suggestion in your previous comment. The problem is that we can have the user as the operation performer, the user being changed or both. So we can't bind a single user object.

wladekb replied:
You can still have a few fluent-ish entry points like `->user($user)` and `->as($user)` and so on but I can see you're not a great supporter of this idea. Then let's leave it as it is.

wladekb commented on lib/Wikia/tests/Service/User/UserPermissionsIntegrationTest.php:
```diff
@@ -0,0 +1,293 @@
+	const TEST_STAFF_USER_NAME = "PermissionTest";
+
+	protected function setUp() {
+		$this->staffUserId = \User::idFromName( self::TEST_STAFF_USER_NAME );
```
How about defining this user as a "fixture' and mock all the dependent data so we don't have to have a working database for running this test?

jcellary replied:
We can - then this won't be an integration test but a unit test. I think both is useful in general.

wladekb replied:
Ok, so let's stick with current implementation

wladekb commented on lib/Wikia/tests/Service/User/UserPermissionsIntegrationTest.php:
```diff
@@ -0,0 +1,293 @@
+
+		$groups = $this->permissionsService()->getExplicitUserGroups( $this->staffUser );
+		$this->assertContains("staff", $groups);
+		$this->assertContains("bureaucrat", $groups);
```
How about some `notContains` (if that's the name) ?

jcellary replied:
good point, I added some

wladekb commented on tests/phpunit/includes/UserTest.php:
```diff
@@ -17,90 +17,14 @@ class UserTest extends MediaWikiTestCase {
 	public function setUp() {
 		parent::setUp();
 
-		$this->savedGroupPermissions = $GLOBALS['wgGroupPermissions'];
```
we don't care about files in `tests/phpunit`

jcellary replied:
What do you mean - they aren't used? So maybe we should just delete them, as otherwise other people like me will waste time on adjusting them.

wladekb replied:
They aren't used. And yes, we should remove them from the codebase.

jcellary replied:
ok deleted whole tests/phpunit folder

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,391 @@
+		return $this->permissionsConfiguration;
+	}
+
+	public function getExplicitGroups(\User $user ) {
```
you could declare
`use User;`
just after the namespace statement at the top of the file and then you can use `User` instead of `\User` in this file.

jcellary replied:
Ok I guess either way is fine (current solution is one line shorter :D ). So do we have a convention that we want to avoid '\' ? If yes I can change that, as I use it quite a bit in other places.

wladekb replied:
I just surveyed one more person (hi @macbre ) and let's stay with the current usage of `\User`

Commit:
Minor fixes based on PR comments

jcellary commented:
@wladekb I made changes according to your comments, let me know if you have any other suggestions

wladekb commented on includes/specials/SpecialUserrights.php:
```diff
@@ -524,17 +553,14 @@ private function groupCheckboxes( $usergroups, $user ) {
 			$disabled = !(
-				( $set && $this->canRemove( $group ) ) ||
-				( !$set && $this->canAdd( $group ) ) );
+				( $set && self::canRemove( $user, $group, $this->isself ) ) ||
```
The canAdd() and canRemove() eventually call PermissionsService::getChangeableGroups() every time which in turn does a heavy array mangling.
We call canAdd()/canRemove() in a loop that goes through all the existing groups, I can see some room for optimizing for reducing cpu waste here

jcellary replied:
You' right I moved the call to getChangeableGroups outside of the loop

wladekb commented on includes/specials/SpecialUserrights.php:
```diff
@@ -21,18 +21,79 @@
+		);
+	}
+
+	public static function isGlobalNonEditableGroup( $globalGroups, $group ) {
```
The `$globalGroups` argument is a static list so we could fetch it ourselves instead of requesting callers to do that in every single place where this method is called

jcellary replied:
That would require this function to not be static as user permissions from the trait is non static and hence can't be accessed. That is why I changed this method to get all required data in parameters

wladekb replied:
`$globalGroups` is part of a global configuration
I'd still prefer the method to take it even if it required sth like:
`Injector::getInjector()->get( PermissionsService::class )->....`
(due to the non-static trait issue you described), I'm assuming that implementation of`Injector::getInjector()->get( ... )` would cache the result and return the object very quickly

jcellary replied:
Paweł told me that the injector is quite slow actually and should be replaced. I'm not sure about caching though.

wladekb commented on includes/specials/SpecialUserrights.php:
```diff
@@ -392,7 +427,7 @@ function switchForm() {
 	 */
 	protected function splitGroups( $groups ) {
-		list( $addable, $removable, $addself, $removeself ) = array_values( $this->changeableGroups() );
+		list( $addable, $removable, $addself, $removeself ) = array_values( $this->getUser()->changeableGroups() );
```
Any reason why you removed the method `changeableGroup()` and inlined its content in a few places in this file?

jcellary replied:
The whole code in changeableGroups in this class was a repetition of logic inside of the user class (which I moved to user permissions service), so I deleted the local copy and replaced it with a call to the central definition.

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsConfiguration.php:
```diff
@@ -0,0 +1,76 @@
+	/**
+	 * Get all the groups who have a given permission
+	 *
+	 * @param $role String Role to check
```
Apparently I don't have any idea what the "role" is. Could you expand the PHPDoc so there is more information about it?

jcellary replied:
I missed that when moving the function - it should be called permission not role (AKA right)

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsService.php:
```diff
@@ -0,0 +1,122 @@
+	 * @param \User $user
+	 * @return string[] Get list of global and local groups to which the given user has been explicitly assigned to.
+	 */
+	public function getExplicitGroups(\User $user );
```
Space mismatch inside parens (no space at the beginning vs one at the end), autoformat this file?

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsService.php:
```diff
@@ -0,0 +1,122 @@
+	 * '*' for all accounts, and autopromoted groups
+	 * @param \User $user
+	 * @param $reCacheAutomaticGroups
+	 * @return Array of String internal group names
```
you've been using "string[]" somewhere above, can we keep one consistent way to specify that?

jcellary replied:
yes

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsService.php:
```diff
@@ -0,0 +1,122 @@
+	 * @param $reCacheAutomaticGroups
+	 * @return Array of String internal group names
+	 */
+	public function getEffectiveGroups(\User $user, $reCacheAutomaticGroups = false );
```
`$recache...` (small c), and the same a few lines below

jcellary replied:
done

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,391 @@
+	private $userPermissions = [];
+
+	public function __construct() {
+		$this->permissionsConfiguration = new PermissionsConfigurationImpl();
```
no injection here?

jcellary replied:
Hmm I guess I could do it, although this is the actual implementation of configuration which ties to the implementation of user permission, so I'm not sure it's necessary here, as this is a pair of implementations.

When I convert it to the new service I'll have a shared data source for both, so I'm sure I'll be then sharing data between them and some internal mechanisms, so direct composition actually seems like the best solution

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,391 @@
+	}
+
+	public function getExplicitGroups(\User $user ) {
+		return array_unique( array_merge (
```
coming back to "calculated every time", for sure `array_unique()` is O(n), I didn't check if that's somewhere called (directly or indirectly) in a loop but if yes that smells like a pure waste of cpu for me, anything we can do about it?

jcellary replied:
I could store it in internal variable, similarly to the groups read from the DB but then I need to remember about purging it when the underlying data changes and I'm not convinced it's worth it. I would totally agree if this was a DB operation etc. but here I'm not convinced it's worth the extra code complexity, especially since in practice these arrays rarely have more then a few elements.

wladekb commented on lib/Wikia/src/Service/User/Permissions/PermissionsServiceImpl.php:
```diff
@@ -0,0 +1,391 @@
+
+	private function loadLocalGroups( $userId ) {
+		if ( !empty( $userId ) && !isset( $this->localExplicitUserGroups[ $userId ] ) ) {
+			$dbr = wfGetDB( DB_MASTER );
```
just wondering if that can also be cached in memcached as the global groups are, any opinion?

jcellary replied:
Possibly. One thing I would keep in mind though is that this code will be deleted in a few weeks and replaced with an invocation of the user permissions service. So I would suggest to leave this implementation as is (as it's tested and working like that for years) and then we I make step 2 of moving it to the service, think about an appropriate caching strategy.

Commit:
Extended unit tests

Commit:
Reformatted code, other minor fixes

Commit:
Merged with dev

Commit:
Removed unused variable

Commit:
Removed possibility for util to manage staff and util groups

Commit:
Added ability for util to remove all groups

Commit:
Merged with dev

jcellary commented:
Note unit tests are failing here, because this branch requires changes in config that are part of this PR: https://github.com/Wikia/config/pull/1620

wladekb commented:
lgtm

macbre commented:
```
1) Wikia\Service\User\Permissions\UserPermissionsIntegrationTest::testShouldReturnExplicitGroups
Failed asserting that an array contains 'bureaucrat'.

/var/lib/jenkins/workspace/unitTests_php_branches/app/lib/Wikia/tests/Service/User/UserPermissionsIntegrationTest.php:54
/var/lib/jenkins/workspace/unitTests_php_branches/app/tests/run-test.php:26

2) Wikia\Service\User\Permissions\UserPermissionsIntegrationTest::testShouldReturnEffectiveGroups
Failed asserting that an array contains 'bureaucrat'.

/var/lib/jenkins/workspace/unitTests_php_branches/app/lib/Wikia/tests/Service/User/UserPermissionsIntegrationTest.php:71
/var/lib/jenkins/workspace/unitTests_php_branches/app/tests/run-test.php:26
```

@jcellary , these changes broke tests suite in SJC (they do pass in Poznań).
